### PR TITLE
Introduced Heterogeneous List.

### PR DESCRIPTION
**Pull request by ghost**
Introducing Heterogeneous List  could be a good step to improve the core library of scala. Sometime it is very useful when dealing with problems that requires TupleN without adding frameworks to a project which often contains a lot of unnecessary tools.

### CODE REVIEW

**Comment by folone** [comment]
Because I'm one of those interested in the `HList` debate, I'll just mention some more interested people to kick off the discussion: @milessabin @szeiger @odersky @larsrh

Also, you might want to keep the [discussion](https://groups.google.com/forum/#!msg/scala-internals/szQ4cHO0lJg/ao6c1RfQ2qcJ) on the mailing list going.

**Comment by soc** [comment]
I really want to see HLists, too, but I think the right approach is not to just add new stuff over time (additionally, HLists don't really fit into the existing collections hierarchy in my opinion).

HLists should replace Scala's TupleX classes and be a drop-in-replacement, so that existing code keeps working, but gains the benefits of HLists over Tuples.

This of course is a much larger project than just adding some new classes, but I would be really interested in getting this done.

So basically I think the naive approach would be to create a branch, add HLists, bootstrap Scala with these additional class available, change the desugaring of `(1, "a", 2.0)` from `Tuple3.apply(1, "a", 2.0)` to `HCons(1, HCons("a", HCons(2.0, HNil)))` (or some `HList3.apply` convenience method), try building it, and then fix every single of those thousand breakages of that change.

**Comment by szeiger** [comment]
@soc HLists have very different performance characteristics from tuples. What you need are HArrays with O(1) indexed access and specialized instances for small arities, e.g. https://github.com/szeiger/ErasedTypes/blob/master/src/test/scala/com/novocode/erased/test/HArrayTest.scala

**Comment by soc** [comment]
@szeiger Yes, absolutely. I just wanted to keep the explanation a bit shorter, because I'm still on the bus.

**Comment by Ichoran** [comment]
@szeiger I agree that HLists have very different performance from tuples.  Are there benchmarking results that show that HArrays are similar to (and, more importantly, competitive with) tuples?  I know that the order will be the same, but constant factors matter rather a lot here.

**Comment by puffnfresh** [comment]
After erasure, I don't see why a HArray would be any different from a tuple class.

**Comment by szeiger** [comment]
I haven't done any performance benchmarks of my implementation. HArrays of low arities (where you care the most about performance) would be pretty much identical to the current tuple classes (including specialization). Higher arities wrap an array of values instead of containing individual Any-typed values. This might cause some overhead, which could be rendered inconsequential by only using this representation for arities > 22.

**Comment by ijuma** [comment]
That sounds like a good approach @szeiger.
